File: home-movie-db-backend/src/main/java/org/thro/sqs/homemoviedb/home_movie_db_backend/web/mapper/MovieMapper.java
package org.thro.sqs.homemoviedb.home_movie_db_backend.web.mapper;

import java.util.ArrayList;
import java.util.List;

import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.thro.sqs.homemoviedb.home_movie_db_backend.business.models.GenreDTO;
import org.thro.sqs.homemoviedb.home_movie_db_backend.business.models.MovieDTO;
import org.thro.sqs.homemoviedb.home_movie_db_backend.web.models.MovieMessage;

@Mapper(componentModel = "spring")
public interface MovieMapper {

    MovieMessage mapToMovieMessage(MovieDTO movie);

    List<MovieMessage> mapToMovieMessage(List<MovieDTO> movies);

    @Mapping(target = "adult", ignore = true)
    @Mapping(target = "originalLanguage", ignore = true)
    @Mapping(target = "originalTitle", ignore = true)
    @Mapping(target = "releaseDate", ignore = true)
    @Mapping(target = "genres", ignore = true)
    MovieDTO mapToMovieDTO(MovieMessage movieMessage);

    default List<String> map(List<GenreDTO> genre){
        if(genre == null){
            return new ArrayList<>();
        }

        return genre.stream().map(GenreDTO::getName).toList();
    }

    List<MovieDTO> mapToMovieDTO(List<MovieMessage> movieMessages);
}

File: home-movie-db-frontend/src/services/mapper/MovieMapper.ts
import { MovieDto, MovieMessage } from "../../models/Movie";

class MovieMapper {
    /**
     * Maps a MovieMessage object to a MovieDto object
     * @param movieMessage The object to be mapped
     * @returns The mapped MovieDto object
     */
    static toMovieDto(movieMessage: MovieMessage): MovieDto {
        return {
            id: movieMessage.id,
            title: movieMessage.title,
            overview: movieMessage.overview,
            genres: movieMessage.genres,
            thumbnailUrl: movieMessage.thumbnailUrl
        };
    }

    
    /**
     * Maps a list of MovieMessage objects to a list of MovieDto objects
     * @param movieMessages The list of objects to be mapped
     * @returns The mapped list of MovieDto objects
     */
    static toMovieDtoList(movieMessages: MovieMessage[]): MovieDto[] {
        return movieMessages.map(movieMessage => MovieMapper.toMovieDto(movieMessage));
    }
}

export default MovieMapper;

